common/djangoapps/student/forms.py

```diff
@@ -213,6 +215,21 @@ def clean_password(self):
                 raise ValidationError(_("Password: ") + "; ".join(err.messages))
         return password
 
+    def clean_email(self):
+        """ Enforce email restrictions (if applicable) """
+        email = self.cleaned_data["email"]
+        if settings.REGISTRATION_EMAIL_PATTERNS_ALLOWED is not None:
+            # This Open edX instance has restrictions on what email addresses are allowed.
+            # We append a '$' to the regexs to prevent the common mistake of using a
+            # pattern like '.*@edx\\.org' which would match 'bob@edx.org.badguy.com'
+            if not any(re.compile(pattern + "$").match(email) for pattern in settings.REGISTRATION_EMAIL_PATTERNS_ALLOWED):
```

Comment:
Good point. I think I forgot about the `re.match()` syntax for a second, since that's what I wanted to do anyways.

Since registration is relatively rare, and Python auto-caches the most recently used regexs anyways, I don't think explicit compiling is a useful optimization here.